components/locale_canonicalizer/src/locale_canonicalizer.rs

```diff
@@ -141,23 +173,62 @@ fn uts35_replacement(
     if ruletype_has_region || (source.id.region.is_none() && replacement.region.is_some()) {
         source.id.region = replacement.region;
     }
-    if let Some(ruletype_variants) = ruletype_variants {
+    if let Some(skips) = ruletype_variants {
         // The rule matches if the ruletype variants are a subset of the source variants.
         // This means ja-Latn-fonipa-hepburn-heploc matches against the rule for
         // hepburn-heploc and is canonicalized to ja-Latn-alalc97-fonipa
-        let mut variants: Vec<subtags::Variant> = source
-            .id
-            .variants
-            .iter()
-            .filter(|x| !ruletype_variants.contains(x))
-            .copied()
-            .collect();
-        for variant in replacement.variants.iter() {
-            variants.push(*variant);
+
+        // We're merging three sorted deduped iterators into a new sequence:
+        // sources - skips + replacements
+
+        let mut sources = source.id.variants.iter().copied().peekable();
+        let mut replacements = replacement.variants.iter().copied().peekable();
+        let mut skips = skips.peekable();
+
+        let mut variants: Vec<Variant> = Vec::new();
+
+        'outer: while let Some(&source) = sources.peek() {
+            // Maybe skip source
+            while let Some(&skip) = skips.peek() {
```

Comment:
Nit (optional): I find it easier to read and reason about nontrivial algorithms like this one if there is a single flat loop, and you manually pull items from each of the iterators. Basically, on each iteration of the loop, you would peek at the 3 iterators, and decide to either skip or write, incrementing the appropriate iterators as necessary. However, what you wrote here is fine and it's not worth changing if it works.